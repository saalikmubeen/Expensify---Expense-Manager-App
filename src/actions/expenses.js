import database from "../firebase";


// add expense action generator
function addExpense(addExpenseObj){
    return {
        type: "ADD_EXPENSE",
        expense: addExpenseObj
    }
}

// remove expense action generator
function removeExpense(payload){
    return {
        type: "REMOVE_EXPENSE",
        id: payload.id
    }
}

// edit expense action generator
function editExpense(id, updates){
    return {
        type: "EDIT_EXPENSE",
        id: id,
        updates: {...updates, amount: Number(updates.amount)}
    }
}


export function asyncAddExpense(addExpenseObj){
    return (dispatch, getState) => {
        var expense = {
            description: addExpenseObj.description,
            note: addExpenseObj.note || "",
            amount: Number(addExpenseObj.amount),
            createdAt: addExpenseObj.createdAt
        }
        database.ref(`users/${getState().user.uid}/expenses`).push(expense).then(function(snapshot){
            dispatch(addExpense({...expense, id: snapshot.key}));
        })
    }
}

function setExpenses(expenses){
    return {
       type: "SET_EXPENSES",
       expenses: expenses
    }
}


export function asyncSetExpenses(){
    return (dispatch, getState) => {
         return database.ref(`users/${getState().user.uid}/expenses`).once("value")
            .then(function(snapshot){
              var expenses = [];

              snapshot.forEach(function(childSnapshot){
                  expenses.push({id: childSnapshot.key, ...childSnapshot.val()})
              })

              dispatch(setExpenses(expenses));
            })
    }
}

export function asyncRemoveExpense(payload){
            return (dispatch, getState) => {
                return database.ref(`users/${getState().user.uid}/expenses/${payload.id}`).remove()
                  .then(function(){
                       dispatch(removeExpense(payload));
                  })
            }
}

export function asyncEditExpense(id, updates){
        return (dispatch, getState) => {
            return database.ref(`users/${getState().user.uid}/expenses/${id}`).update({
                ...updates, amount: Number(updates.amount)
            }).then(function(){
                dispatch(editExpense(id, updates));
            })
        }
}

// ============================================================================================================

// action generators always have to return an object containing type property
// that gets dispatched to update the store in some way
// But some times we may have to do some other task when updating store like make ajax request
// basic pattern to call/dispatch (async) function inside dispatch()
// this function returns another function with dispatch as argument which in turn dispatches the original action generator.
// this is possible only if we use middleware like redux-thunk