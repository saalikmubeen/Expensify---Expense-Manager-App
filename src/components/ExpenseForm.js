import React, { Component } from 'react'
import moment from "moment";
import 'react-dates/initialize'
import {SingleDatePicker} from "react-dates";
import 'react-dates/lib/css/_datepicker.css';

class ExpenseForm extends Component {
    constructor(props){
        super(props);
        this.state = {
            description: this.props.expense ? this.props.expense.description : "",
            note: this.props.expense ? this.props.expense.note : "",
            amount: this.props.expense ? this.props.expense.amount : "",
            createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
            calenderFocused: false,
            error: ""
        }
    }

    handleChange = (e) => {
        this.setState({[e.target.name]: e.target.value});
    }

    handleSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount){
            return this.setState({error: "Provide a description and amount!"})
        }

        this.setState({error: ""});
        this.props.saveExpense({
             description: this.state.description, 
             note: this.state.note, 
             amount: this.state.amount, 
             createdAt: this.state.createdAt.valueOf()})
    }

    render() {
        return (
            
            <form className="form" onSubmit={this.handleSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input
                    type="text"
                    placeholder="Description"
                    name="description"        
                    autoFocus
                    autoComplete="off"
                    className="text-input"
                    value={this.state.description}
                    onChange={this.handleChange}
                />
                
                <input
                    type="text"
                    placeholder="Amount"
                    name="amount"
                    autoComplete="off"
                    className="text-input"
                    value={this.state.amount}
                    onChange={this.handleChange}
                />

                <SingleDatePicker
                    date={this.state.createdAt} // momentPropTypes.momentObj or null
                    onDateChange={date => this.setState({ createdAt: date })} // PropTypes.func.isRequired
                    focused={this.state.calenderFocused} // PropTypes.bool
                    onFocusChange={({ focused }) => this.setState({ calenderFocused: focused })} // PropTypes.func.isRequired
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />

                <textarea
                    placeholder="Add a note for your expense (optional)"
                    name="note"
                    className="textarea"
                    value={this.state.note}
                    onChange={this.handleChange}
                >
                </textarea>

                <div>
                    <button className="button">Save Expense</button>
                </div>
            </form>
            
        )
    }
}


export default ExpenseForm;