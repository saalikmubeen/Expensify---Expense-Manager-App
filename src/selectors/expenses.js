import moment from "moment";

// a function just to arrange and show the expenses based on the current filters applied.
function getVisibleExpenses(expenses, filters){
    return expenses.filter((expense) => {
        var textMatch = expense.description.toLowerCase().includes(filters.text.toLowerCase());

        var startDateMatch = filters.startDate ? moment(expense.createdAt).isSameOrAfter(filters.startDate, 'day') : true;
    

        var endDateMatch = filters.endDate ? moment(expense.createdAt).isSameOrBefore(filters.endDate, 'day') : true;

        return textMatch && startDateMatch && endDateMatch; 
    }).sort((a, b) => {
        if(filters.sortBy === "date"){
            return a.createdAt < b.createdAt ? 1 : -1
        }else if(filters.sortBy === "amount"){
            return a.amount < b.amount ? 1 : -1
        }
    })
};

export default getVisibleExpenses;





// function getVisibleExpenses(expenses, filters){
//     return expenses.filter((expense) => {
//         var textMatch = expense.description.toLowerCase().includes(filters.text.toLowerCase());
//         var startDateMatch;
//         if(typeof expense.startDate === "number" && typeof filters.startDate === "number"){
//             startDateMatch = expense.startDate >= filters.startDate;
//         }else{
//             startDateMatch = true
//         }

//         var endDateMatch;
//         if(typeof expense.endDate === "number" && typeof filters.endDate === "number"){
//           endDateMatch = expense.endDate <= filters.endDate;
//       }else{
//           endDateMatch = true
//       }

//         return textMatch && startDateMatch && endDateMatch; 
//     }).sort((a, b) => {
//         if(filters.sortBy === "date"){
//             return a.createdAt < b.createdAt ? 1 : -1
//         }else if(filters.sortBy === "amount"){
//             return a.amount < b.amount ? 1 : -1
//         }
//     })
// };
