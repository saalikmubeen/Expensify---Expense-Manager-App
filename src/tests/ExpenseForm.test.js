import React from "react";
import { shallow } from "enzyme";
import moment from "moment";
import ExpenseForm from "../components/ExpenseForm";
import expenses from "./expenses";

test("Expect to render ExpenseForm without expense data", () => {
    const wrapper = shallow(<ExpenseForm />);
    // console.log(wrapper.length);

    expect(wrapper).toMatchSnapshot();
});

test("Expect to render ExpenseForm with expense data", () => {
    const wrapper = shallow(<ExpenseForm expense={expenses[2]} />);
    // console.log(wrapper.length);

    expect(wrapper).toMatchSnapshot();
});

test("Expect form submission without description or without amount", () => {
    const wrapper = shallow(<ExpenseForm />);

    expect(wrapper).toMatchSnapshot();

    wrapper.find("form").simulate("submit", {
        preventDefault: () => {},
    }); // second argument passed to simulate is the event object onSubmit expects

    expect(wrapper.state("error")).toBe("Provide a description and amount!");
    expect(wrapper.state("error").length).toBeGreaterThan(0);

    expect(wrapper).toMatchSnapshot();
});

test("Expect form submission with valid form submission", () => {
    const saveExpenseSpy = jest.fn();
    const wrapper = shallow(
        <ExpenseForm expense={expenses[1]} saveExpense={saveExpenseSpy} />
    );

    wrapper.find("form").simulate("submit", {
        preventDefault: () => {},
    });

    expect(wrapper.state("error")).toBe("");
    expect(saveExpenseSpy).toHaveBeenLastCalledWith({
        description: expenses[1].description,
        note: expenses[1].note,
        amount: expenses[1].amount,
        createdAt: expenses[1].createdAt,
    });

    expect(wrapper).toMatchSnapshot();
});

test("Should set description on input change", () => {
    const wrapper = shallow(<ExpenseForm />);

    wrapper
        .find("input")
        .at(0)
        .simulate("change", {
            target: { name: "description", value: "hi" },
        });

    expect(wrapper.state("description")).toBe("hi");

    expect(wrapper).toMatchSnapshot();
});

test("Should set note on input change", () => {
    const wrapper = shallow(<ExpenseForm />);

    wrapper.find("textarea").simulate("change", {
        target: { name: "note", value: "note it" },
    });

    expect(wrapper.state("note")).toBe("note it");

    expect(wrapper).toMatchSnapshot();
});

test("Should set new Date on date change", () => {
    const now = moment();
    const wrapper = shallow(<ExpenseForm />);

    wrapper.find("#datePicker").prop("onDateChange")(now);

    expect(wrapper.state("createdAt")).toEqual(now);

    expect(wrapper).toMatchSnapshot();
});
