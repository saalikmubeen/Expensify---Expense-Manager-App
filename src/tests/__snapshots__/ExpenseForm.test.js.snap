// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect form submission with valid form submission 1`] = `ShallowWrapper {}`;

exports[`Expect form submission without description or without amount 1`] = `ShallowWrapper {}`;

exports[`Expect form submission without description or without amount 2`] = `ShallowWrapper {}`;

exports[`Expect to render ExpenseForm with expense data 1`] = `ShallowWrapper {}`;

exports[`Expect to render ExpenseForm without expense data 1`] = `ShallowWrapper {}`;

exports[`Should set description on input change 1`] = `ShallowWrapper {}`;

exports[`Should set new Date on date change 1`] = `ShallowWrapper {}`;

exports[`Should set note on input change 1`] = `ShallowWrapper {}`;
